tools/marketplace/helper/bundleHelper.go

```diff
@@ -0,0 +1,311 @@
+	CRDs     string `yaml:"customResourceDefinitions"`
+	Packages string `yaml:"packages"`
+}
+
+//Bundle - OLM bundle
+type Bundle struct {
+	Data Data `yaml:"data"`
+}
+
+//BundleHelper - object that provides logic of fetching OLM bundle from marketplace quay repo
+type BundleHelper struct {
+	repo      string
+	namespace string
+	Pkgs      []Pkg
+	CRDs      []v1beta1.CustomResourceDefinition
+	CSVs      []yaml.MapSlice
+}
+
+//NewBundleHelper - Ctor
+func NewBundleHelper(repo, namespace string) (*BundleHelper, error) {
+	bh := &BundleHelper{
+		repo:      repo,
+		namespace: namespace,
+	}
+	if err := bh.downloadAndParseBundle(); err != nil {
+		klog.Errorf("Failed to  download bundle", err)
+		return nil, err
+	}
+	return bh, nil
+}
+
+func (bh *BundleHelper) downloadAndParseBundle() error {
+
+	// prepare the app registry client
+	options := appregistry.Options{
+		Source:    "https://quay.io/cnr",
+		AuthToken: "", // not needed for public applications
+	}
+	client, err := appregistry.NewClientFactory().New(options)
+	if err != nil {
+		klog.Errorf("Failed to  create new appregistry client", err)
+		return err
+	}
+
+	// get latest bundle info
+	bundles, err := client.ListPackages(bh.namespace)
+	if err != nil {
+		klog.Errorf("Failed to  list packages", err)
+		return err
+	}
+
+	if len(bundles) == 0 {
+		klog.Infof("no new bundles")
+		return nil
+	}
+	cdiBundles, err := bh.getRepoBundles(bundles)
+	if err != nil {
+		klog.Errorf("Failed to  get repo bundles", err)
+		return err
+	}
+
+	if len(cdiBundles) == 0 || cdiBundles[0] == nil {
+		klog.Infof("no new bundles")
+		return nil
+	}
+
+	bundleMetaData := cdiBundles[0]
+
+	// download bundle
+	data, err := client.RetrieveOne(bundleMetaData.ID(), bundleMetaData.Release)
+	if err != nil {
+		klog.Errorf("Failed to retriev bundle data", err)
+		return err
+	}
+
+	// parse bundle into package, CRDs and CSVs
+	bundle := Bundle{}
+	if err := yaml.Unmarshal(data.RawYAML, &bundle); err != nil {
+		klog.Errorf("Failed to unmarshal bundle data", err)
+		return err
+	}
+
+	if err := yaml.Unmarshal([]byte(bundle.Data.Packages), &bh.Pkgs); err != nil {
+		klog.Errorf("Failed to unmarshal package data", err)
+		return err
+	}
+
+	if err := yaml.Unmarshal([]byte(bundle.Data.CSVs), &bh.CSVs); err != nil {
+		klog.Errorf("Failed to unmarshal CSV data", err)
+		return err
+	}
+
+	// use k8s json unmarshaller for CRDs for filling metadata correctly
+	crds, err := yaml2.YAMLToJSON([]byte(bundle.Data.CRDs))
+	if err != nil {
+		klog.Errorf("Failed to convert CRD data to json format", err)
+		return err
+	}
+	if err := json.Unmarshal([]byte(crds), &bh.CRDs); err != nil {
+		klog.Errorf("Failed to unmarshal CRD data", err)
+		return err
+	}
+
+	return nil
+}
+
+func (bh *BundleHelper) getRepoBundles(bundles []*datastore.RegistryMetadata) ([]*datastore.RegistryMetadata, error) {
+	list := make([]*datastore.RegistryMetadata, len(bundles))
+	i := 0
+	for j := 0; j < len(bundles); j++ {
+		bundle := bundles[j]
```

Comment:
this can be done easier:
`for _, bundle := range bundles {...}`
replace `_` with `j` if you need it

I would also start with an empty `list` above, and later on just `list = append(list, metadata)`, no need for `i`

Reply:
Done